### PR TITLE
Use Type System Package

### PR DESCRIPTION
**Pull request by Alfred-Mountfield**
## 🌟 What is the purpose of this PR?

The definition of the Block Protocol types are being implemented over on the block protocol repo: https://github.com/blockprotocol/blockprotocol/pull/503

This PR starts to migrate the structs to use the crate there

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202775849963301/f) _(internal)_

## 🔍 What does this change?

- Adds a git-resolved dependency on the type-system crate
- Deletes all duplicate/redundant code
- Modifies implementations to match new signatures
- Updates test data to pass new validation logic

## 📜 Does this require a change to the docs?

- Probably not, and regenerating the OpenAPI spec didn't result in a diff

## ⚠️ Known issues

`ValidationError` on the type-system package does not have all of the variants we have currently in `dev/graph`. This PR maintains a duplicate `ValidationError`, but there was one spot where this caused problems I couldn't find a nice solution around `hash_graph/lib/graph/src/ontology/types/entity_type/mod.rs:130`

## 🐾 Next steps

- Continue to migrate as the Type System package continues to get developed to support Entity Types and Link Types

## 🛡 What tests cover this?

- These are core primitives so the full test suite interfaces with this

## ❓ How to test this?

Follow the README

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/ontology/types/entity_type/mod.rs [hunk 1]
```diff
@@ -38,15 +33,15 @@ impl EntityTypeReference {
...
+        } else {
+            Err(ValidationError::BaseUriMismatch {
                 base_uri: base_uri.clone(),
-                versioned_uri: self.uri().clone()
-            }
-        );
-        Ok(())
+                versioned_uri: self.uri().clone(),
+            })
+        }
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/ontology/types/entity_type/mod.rs:44]: What is the reason for this change?

File: packages/graph/hash_graph/lib/graph/src/ontology/types/entity_type/links.rs [hunk 2]
```diff
@@ -93,9 +92,9 @@ pub struct Links {
 }
 
 impl TryFrom<LinksRepr> for Links {
-    type Error = ValidationError;
+    type Error = crate::ontology::types::error::ValidationError;
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/ontology/types/entity_type/links.rs:95]: I guess this will also use `type_system::ValidationError` as soon as `LinkType`s were migrated?

File: packages/graph/hash_graph/lib/graph/src/ontology/types/link_type/mod.rs [hunk 3]
```diff
@@ -67,7 +66,7 @@ mod tests {
     fn test_link_type_schema(schema: &serde_json::Value) -> LinkType {
         let link_type: LinkType = serde_json::from_value(schema.clone()).expect("invalid schema");
         assert_eq!(
-            serde_json::to_value(link_type.clone()).expect("could not serialize"),
+            serde_json::to_value(link_type.clone()).expect("Could not serialize"),
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/ontology/types/link_type/mod.rs:69]: Error messages should not be capitalized.
```suggestion
            serde_json::to_value(link_type.clone()).expect("could not serialize"),
```

Could you also please look at other locations you changed?

File: packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs [hunk 4]
```diff
@@ -73,7 +73,7 @@ where
                         WHERE base_uri = $1
                     );
                 "#,
-                &[&base_uri],
+                &[&base_uri.to_string()],
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs:76]: Not really happy we need to create a string only for passing it to the database. We should probably add a function like `BaseUri::as_str()` in a follow-up.

File: packages/graph/hash_graph/tests/test_data/entity/book.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
-  "https://blockprotocol.org/@alice/types/property-type/name": "The Time Machine",
+  "https://blockprotocol.org/@alice/types/property-type/name/": "The Time Machine",
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/tests/test_data/entity/book.json:2]: Having the `BaseURI` like this looks weird to me, we should allow both ways.

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

File: packages/graph/hash_graph/lib/graph/src/ontology/types/entity_type/mod.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/ontology/types/entity_type/mod.rs:44]: `ValidateUri` is defined inside the Type System Package, which does not use `error-stack` currently.

File: packages/graph/hash_graph/lib/graph/src/ontology/types/link_type/mod.rs (hunk 3 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/ontology/types/link_type/mod.rs:69]: We're discussing this offline and struggling to find guidelines or conventions for panics vs errors. Likely going to go with lower case

File: packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs (hunk 4 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs:76]: I'm also happy for us to block merging of this until we've done a few changes on the upstream crate

File: packages/graph/hash_graph/lib/graph/src/ontology/types/entity_type/mod.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/ontology/types/entity_type/mod.rs:44]: Gotcha, didn't noticed the `std::result::Result`, thanks for clarifying

File: packages/graph/hash_graph/tests/test_data/entity/book.json (hunk 5 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/tests/test_data/entity/book.json:2]: We discussed this offline and we need some further changes in the upstream crate.
Summary: we need (and require) the trailing slash.

File: packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs (hunk 4 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs]: These should be updated, I don't _think_ I've missed any

File: packages/graph/hash_graph/lib/graph/src/ontology/types/link_type/mod.rs (hunk 3 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/ontology/types/link_type/mod.rs]: Have found and replaced all in the project, if there are some expects with the string on a newline I might have missed them

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs [hunk 6]
```diff
@@ -108,13 +111,13 @@ async fn by_link_type_by_source_entity_id(
...
             "#,
-            &[&source_entity_id, link_type_uri, &i64::from(link_type_version)],
+            &[&source_entity_id, &link_type_base_uri.as_str(), &i64::from(link_type_version)],
         )
         .await
         .into_report()
         .change_context(QueryError)
         .attach_printable(source_entity_id)
-        .attach_printable(link_type_uri.clone())?;
+        .attach_printable(link_type_base_uri.clone())?;
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:120]:
```suggestion
        .attach_printable_lazy(|| link_type_base_uri.clone())?;
```
Otherwise, it would always be cloned, regardless if it's an error or not

**Review by TimDiekmann — COMMENTED** [review]
LGTM 🎉 

Found one issue (I know you didn't introduce it), but could you fix it as a drive-by?

**Review by TimDiekmann — APPROVED** [review]
